### PR TITLE
[feature] Multiple emails

### PR DESCRIPTION
**Pull request by lyndsysimon**
Status: ready for review

### CODE REVIEW

**Comment by lbanner** [comment]
@lyndsysimon Let's focus this PR on the limited multiple emails requirement i.e. ability to add emails to a user account that are non-existent, or unregistered/unconfirmed -- and then knock this out -- OK

File: framework/auth/__init__.py [hunk 1]
```diff
@@ -117,20 +117,17 @@ def verify_two_factor(user_id, two_factor_code):
     return response
 
 
-def login(username, password):
+def login(email, password):
```
Comment (sloria) [inline framework/auth/__init__.py]: Just curious: why was this changed? The field name is called `username` so it sees intuitive for the parameter name to be the same.

Comment (sloria) [inline framework/auth/__init__.py]: You don't need to change it back, though; just wondering what the rationale was.

File: framework/auth/core.py [hunk 2]
```diff
@@ -541,14 +582,35 @@ def _set_email_token_expiration(self, token, expiration=None):
         self.email_verifications[token]['expiration'] = expiration
         return expiration
 
-    def add_email_verification(self, email, expiration=None):
+    def add_unconfirmed_email(self, email, expiration=None):
         """Add an email verification token for a given email."""
+        # TODO: If the unconfirmed email is already on the User, refresh the token
+
+        # verify email
+        if not re.match(r'^.+@[^.].*\.[a-z]{2,10}$', email, re.IGNORECASE):
```
Comment (sloria) [inline framework/auth/core.py]: This should be in a separate validator callable.

File: framework/auth/core.py [hunk 3]
```diff
@@ -541,14 +582,35 @@ def _set_email_token_expiration(self, token, expiration=None):
...
+            raise ValueError("Invalid Email")
+
         token = generate_confirm_token()
 
         self.email_verifications[token] = {'email': email.lower()}
         self._set_email_token_expiration(token, expiration=expiration)
         return token
 
+    def remove_unconfirmed_emails(self, *args):
```
Comment (sloria) [inline framework/auth/core.py]: In case we need to add extra arguments, it might be more future-proof to pass `emails` as a positional argument.

File: framework/auth/core.py [hunk 4]
```diff
@@ -858,15 +950,76 @@ def get_daily_digest_log_ids(self):
...
-    def merge_user(self, user, save=False):
+    @property
+    def can_be_merged(self):
+        """The ability of the `merge_user` method to fully merge the user"""
+        return (
+            self.is_confirmed is False and
+            self.get_addons() == []
+        )
+
+    def merge_user(self, user):
```
Comment (sloria) [inline framework/auth/core.py]: Why was the `save` parameter removed?

File: framework/auth/core.py [hunk 5]
```diff
@@ -858,15 +950,76 @@ def get_daily_digest_log_ids(self):
...
+        # - unclaimed records should be connected to only one user
+        user.unclaimed_records = {}
+
+        security_messages = user.security_messages.copy()
+        security_messages.update(self.security_messages)
+        self.security_messages = security_messages
+
+        for key, value in user.mailing_lists.iteritems():
+            # subscribe to each list if either user was subscribed
+            self.mailing_lists[key] = self.mailing_lists[key] or value
```
Comment (sloria) [inline framework/auth/core.py]: Could a `KeyError` be raised here?

File: framework/auth/views.py [hunk 6]
```diff
@@ -193,23 +193,40 @@ def confirm_email_get(**kwargs):
...
-    # Return data for the error template
-    return {
-        'code': http.BAD_REQUEST,
-        'message_short': 'Link Expired',
-        'message_long': language.LINK_EXPIRED
-    }, http.BAD_REQUEST
+    try:
+        user.confirm_email(token)
+    except exceptions.EmailConfirmTokenError as e:
+        return {
```
Comment (sloria) [inline framework/auth/views.py]: Why not `raise HTTPError(http.BAD_REQUEST, data=dict(message_short=e.message_short...))`?

File: website/static/js/accountSettings.js [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+var $osf = require('js/osfHelpers.js');
+var ko = require('knockout');
+var oop = require('js/oop.js');
+var Raven = require('raven-js');
+
+require('knockout.punches');
+ko.punches.enableAll();
+
+
+var userEmail = oop.defclass({
```
Comment (sloria) [inline website/static/js/accountSettings.js]: Use `UpperCamelcase` for constructors.

File: website/static/js/accountSettings.js [hunk 8]
```diff
@@ -0,0 +1,170 @@
+'use strict';
+
+var $ = require('jquery');
+var $osf = require('js/osfHelpers.js');
```
Comment (sloria) [inline website/static/js/accountSettings.js]: Omit the `.js` extension.

File: website/static/js/accountSettings.js [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        var self = this;
+        var profile = data.profile;
+
+        this.id(profile.id);
+        self.emails([]);
+        ko.utils.arrayMap(profile.emails, function(each) {
+            self.emails.push(new userEmail(each));
+        });
+    },
+    fetchData: function (callback) {
```
Comment (sloria) [inline website/static/js/accountSettings.js]: If you return `request`--which is a promise--you no longer need the `callback` parameter.

See https://cosdev.readthedocs.org/en/latest/style_guides/javascript.html#promises

Comment (sloria) [inline website/static/js/accountSettings.js]: It may make sense to separate data fetching from the ViewModel. Perhaps create a `ProfileClient` that does the work of fetching email info from the server.

File: website/static/js/accountSettings.js [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        request.fail(function(xhr, status, error) {
+            $osf.growl('Error', 'Could not fetch user profile.', 'danger');
+            Raven.captureMessage('Error fetching user profile', {
+                url: url,
+                status: status,
+                error: error
+            });
+        });
+    },
+    pushUpdates: function (callback) {
```
Comment (sloria) [inline website/static/js/accountSettings.js]: See note above re: promises over callbacks.

File: website/static/js/pages/profile-account-settings-page.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+var $ = require('jquery');
+var $osf = require('js/osfHelpers.js');
```
Comment (sloria) [inline website/static/js/pages/profile-account-settings-page.js]: Same here; omit `.js`.

File: website/templates/profile/account.mako [hunk 12]
```diff
@@ -5,39 +5,101 @@
...
+                            <div class="form-group">
+                                <label for="old-password">Old password</label>
+                                <input type="password" class="form-control" name="old_password">
+                            </div>
+                            <div class="form-group">
+                                <label for="password">New password</label>
+                                <input type="password" class="form-control" name="new_password">
+                            </div>
+                            <div class="form-group">
+                                <label for="password2">Confirm new password</label>
```
Comment (sloria) [inline website/templates/profile/account.mako]: Since you're providing the `for` attribute, it would make sense to add the `name` attribute to the `input` tags.

Comment (sloria) [inline website/templates/profile/account.mako]: Oh, I just saw that you are providing `name`, but the `name` attributes are inconsistent with the `for` attributes.

File: website/static/js/accountSettings.js [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+    }
+});
+
+
+var userProfile = oop.defclass({
+    constructor: function(urls) {
+        this.urls = urls;
+
+        // TODO: Pass these in instead!
+        this.urls = {
```
Comment (sloria) [inline website/static/js/accountSettings.js]: If you separate the CRUD operations into the `ProfileClient` suggested in my previous comment, you won't need to pass these in; they can just be internal data for the `ProfileClient`.

File: scripts/migration/migrate_unconfirmed_user_emails.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        Q('date_confirmed', 'eq', None)
+        & Q('emails', 'ne', [])
+    )
+
+
+def remove_unconfirmed_emails(user):
+    user.emails = []
+
+
+class TestMigrateUnconfirmedEmails(OsfTestCase):
```
Comment (sloria) [inline scripts/migration/migrate_unconfirmed_user_emails.py]: We no longer put tests in the migration script (to prevent `tests.base` from, e.g. silencing loggers). These should go in `scripts/tests`.

I've just updated COSDev with this new guideline: https://cosdev.readthedocs.org/en/latest/osf/migrations.html#migrations

**Comment by sloria** [comment]
1st pass finished. :point_up:

File: framework/auth/__init__.py (hunk 1 above)
Comment (lyndsysimon) [inline framework/auth/__init__.py]: The purpose of the `username` field has changed - it's now the primary email for the account. It's possible for the primary email to be unconfirmed, as is the case for an unregistered user or a user who has not yet completed email verification.

File: framework/auth/core.py (hunk 3 above)
Comment (lyndsysimon) [inline framework/auth/core.py]: I don't have a strong feeling about it, but it was intentional - as long as extra arguments are keywords, it shouldn't be a problem.

File: framework/auth/core.py (hunk 4 above)
Comment (lyndsysimon) [inline framework/auth/core.py]: It wasn't used anywhere, and the user being merged (vs. the one being merged into) was always saved, regardless of the value passed.

File: framework/auth/core.py (hunk 5 above)
Comment (lyndsysimon) [inline framework/auth/core.py]: Good catch, yes.

File: framework/auth/views.py (hunk 6 above)
Comment (lyndsysimon) [inline framework/auth/views.py]: I agree. I was following the convention of the existing code so as not to have to think about the impact of refactoring it. There's shouldn't be an impact to this change though, so I've made it.

File: website/static/js/accountSettings.js (hunk 7 above)
Comment (lyndsysimon) [inline website/static/js/accountSettings.js]: Done.

File: website/static/js/accountSettings.js (hunk 9 above)
Comment (lyndsysimon) [inline website/static/js/accountSettings.js]: Returning `request` is of course correct - I'm not sure why I didn't do that in the first place, honestly.

File: website/templates/profile/account.mako (hunk 12 above)
Comment (lyndsysimon) [inline website/templates/profile/account.mako]: I didn't change that code, only moved it. No worries though, I've made this change.

File: scripts/migration/migrate_unconfirmed_user_emails.py (hunk 14 above)
Comment (lyndsysimon) [inline scripts/migration/migrate_unconfirmed_user_emails.py]: done

File: website/static/js/accountSettings.js (hunk 8 above)
Comment (lyndsysimon) [inline website/static/js/accountSettings.js]: done

File: website/static/js/accountSettings.js (hunk 10 above)
Comment (lyndsysimon) [inline website/static/js/accountSettings.js]: done

File: website/static/js/pages/profile-account-settings-page.js (hunk 11 above)
Comment (lyndsysimon) [inline website/static/js/pages/profile-account-settings-page.js]: done

**Comment by lyndsysimon** [comment]
Ready for second pass.

File: framework/auth/__init__.py (hunk 1 above)
Comment (sloria) [inline framework/auth/__init__.py]: :+1:

File: framework/auth/core.py (hunk 3 above)
Comment (sloria) [inline framework/auth/core.py]: True, but then you couldn't the keyword arguments in the method definition. You'd have to do:

```python
def remove_unconfirmed_emails(self, save=False , *args):

# or

def remove_unconfirmed_emails(self, *args, **kwargs):
    save = kwargs.get('save', False)
```

Neither is ideal.

File: tests/test_models_user.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
+from nose.tools import *
+
+from framework import auth
+from framework.auth import exceptions
+from website import models
+from tests import base
+from tests import factories
+
+
+class UserTestCase(base.OsfTestCase):
```
Comment (sloria) [inline tests/test_models_user.py]: We might want want to move our existing `User` model tests into this file. Also, I like the idea of breaking up our test modules by model like you've done here. Maybe use a new directory: `tests/test_models/test_user.py`.

This can be deferred.

File: website/profile/views.py [hunk 16]
```diff
@@ -85,20 +86,64 @@ def update_user(auth):
...
-    if 'timezone' in data:
-        if data['timezone']:
-            user.timezone = data['timezone']
+
+    if 'emails' in data:
+        # removals
+        removed_emails = [
+            each
+            for each in user.emails + user.unconfirmed_emails
+            if each not in [x['address'] for x in data['emails']]
```
Comment (sloria) [inline website/profile/views.py]: Should `each` and `x['address']` be lowercased for case-insensitivity?

File: website/profile/views.py [hunk 17]
```diff
@@ -85,20 +86,64 @@ def update_user(auth):
...
+            if address in user.emails:
+                user.emails.remove(address)
+            user.remove_unconfirmed_emails(address)
+
+        # additions
+        added_emails = [
+            each['address']
+            for each in data['emails']
+            if each['address'] not in user.emails
+            and each['address'] not in user.unconfirmed_emails
```
Comment (sloria) [inline website/profile/views.py]: Same here re: case-insensitivity

File: website/profile/views.py [hunk 18]
```diff
@@ -85,20 +86,64 @@ def update_user(auth):
...
+        ]
+
+        for address in added_emails:
+            try:
+                user.add_unconfirmed_email(address)
+            except ValueError:
```
Comment (sloria) [inline website/profile/views.py]: I'm not seeing where `add_confirmed_email` can raise a `ValueError`. Can you please clarify?

File: website/profile/views.py [hunk 19]
```diff
@@ -85,20 +86,64 @@ def update_user(auth):
...
+                continue
+
+            # TODO: This setting is now named incorrectly.
+            if settings.CONFIRM_REGISTRATIONS_BY_EMAIL:
+                send_confirm_email(user, email=address)
+
+        # set username
+        username = None
+        try:
+            username = [each for each in data['emails']
```
Comment (sloria) [inline website/profile/views.py]: This could raise a `KeyError` if the client sends, e.g. `[{'primary': True}, ...]`.

Comment (sloria) [inline website/profile/views.py]: Instead of building the entire list in-memory, you can call `next` on a generator to return the first matching element. You can also pass a default value as the second argument.

```python
username = next(
    (each for each in data.get('emails', [])
        if each.get('primary')), None
)
```

File: website/profile/views.py [hunk 20]
```diff
@@ -85,20 +86,64 @@ def update_user(auth):
...
+        # make sure the new username has already been confirmed
+        if username and username in user.emails:
+            user.username = username
+
     if 'locale' in data:
         if data['locale']:
             locale = data['locale'].replace('-', '_')
             user.locale = locale
+    if 'timezone' in data:
+        if data['timezone']:
```
Comment (sloria) [inline website/profile/views.py]: Could you just do `if data.get('timezone', None):`?

File: website/static/js/accountSettings.js [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+});
+
+
+var UserProfile = oop.defclass({
+    constructor: function () {
+
+        this.id = ko.observable();
+        this.emails = ko.observableArray();
+
+        this.primaryEmail = ko.computed(function () {
```
Comment (sloria) [inline website/static/js/accountSettings.js]: I believe this can be a `pureComputed`.

File: website/static/js/accountSettings.js [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+            return retval;
+        }.bind(this));
+
+    }
+});
+
+
+var UserProfileClient = oop.defclass({
+    constructor: function () {
+        this.urls = {
```
Comment (sloria) [inline website/static/js/accountSettings.js]: `urls` can probably be a class variable; it won't vary between instances.

File: website/static/js/accountSettings.js [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+            ret.resolve(profile);
+
+        }.bind(this));
+        request.fail(function(xhr, status, error) {
+            $osf.growl('Error', 'Could not fetch user profile.', 'danger');
+            Raven.captureMessage('Error fetching user profile', {
+                url: this.urls.fetch,
+                status: status,
+                error: error
+            });
```
Comment (sloria) [inline website/static/js/accountSettings.js]: Would make sense to reject the promise here.

```js
ret.reject(xhr, status, error);
```

File: website/static/js/accountSettings.js [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+        }.bind(this));
+
+        return ret.promise();
+    },
+    update: function (profile) {
+        var request = $osf.putJSON(
+            this.urls.update,
+            this.serialize(profile)
```
Comment (sloria) [inline website/static/js/accountSettings.js]: The `success` parameter is deprecated: https://github.com/lyndsysimon/osf.io/blob/e8774f092a2941a79e7360e24c88495515220a33/website/static/js/osfHelpers.js#L29-30 .

Just add another `.done` or pass an array of functions to the first `.done`.

File: website/static/js/accountSettings.js [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+        ).done(function (data) {
+            profile = this.unserialize(data, profile);
+        }.bind(this)).fail(function(xhr, status, error) {
```
Comment (sloria) [inline website/static/js/accountSettings.js]: I find it more clear and concise to do `var self = this;` rather than having to bind everywhere.

Comment (sloria) [inline website/static/js/accountSettings.js]: https://cosdev.readthedocs.org/en/latest/style_guides/javascript.html#style

File: framework/auth/core.py (hunk 4 above)
Comment (sloria) [inline framework/auth/core.py]: The `save` parameter applied to the user being merged into (`self`). We should avoid implicit saves whenever possible.

File: framework/auth/core.py [hunk 26]
```diff
@@ -605,26 +675,51 @@ def verify_claim_token(self, token, project_id):
...
+            self.register(self.username)
+            self.date_confirmed = dt.datetime.utcnow()
+        # Revoke token
+        del self.email_verifications[token]
+
+        # TODO: We can't assume that all unclaimed records are now claimed.
+        # Clear unclaimed records, so user's name shows up correctly on
+        # all projects
+        self.unclaimed_records = {}
+        self.save()
```
Comment (sloria) [inline framework/auth/core.py]: Same here. Avoid implicit saves.

File: framework/auth/core.py [hunk 27]
```diff
@@ -605,26 +675,51 @@ def verify_claim_token(self, token, project_id):
...
+
+        # We must manually update search here because fullname wasn't changed
+        self.update_search()
```
Comment (sloria) [inline framework/auth/core.py]: Could we add `'date_confirmed'` to `SEARCH_UPDATE_FIELDS` and remove this manual update?

File: framework/auth/core.py [hunk 28]
```diff
@@ -605,26 +675,51 @@ def verify_claim_token(self, token, project_id):
...
+        """Confirm the email address associated with the token"""
+        email = self._get_unconfirmed_email_for_token(token)
+
+        # If this email is confirmed on another account, abort
+        if User.find(Q('emails', 'eq', email)).count() > 0:
+            raise exceptions.DuplicateEmailError()
+
+        # If another user has this email as its username, get it
+        try:
+            unregistered_user = User.find_one(Q('username', 'eq', email) &
```
Comment (sloria) [inline framework/auth/core.py]: Might need to use `iexact` for the query for case-insensitivity.

File: framework/auth/core.py [hunk 29]
```diff
@@ -605,26 +675,51 @@ def verify_claim_token(self, token, project_id):
...
-            self.update_search()
-            self.update_search_nodes()
-            return True
-        else:
-            return False
+        """Confirm the email address associated with the token"""
+        email = self._get_unconfirmed_email_for_token(token)
+
+        # If this email is confirmed on another account, abort
+        if User.find(Q('emails', 'eq', email)).count() > 0:
```
Comment (sloria) [inline framework/auth/core.py]: Might need to account for case-insensitivity here as well.

File: website/static/js/accountSettings.js [hunk 30]
```diff
@@ -0,0 +1,187 @@
...
+var ko = require('knockout');
+var oop = require('js/oop');
+var Raven = require('raven-js');
+
+require('knockout.punches');
+ko.punches.enableAll();
+
+
+var UserEmail = oop.defclass({
+    constructor: function() {
```
Comment (sloria) [inline website/static/js/accountSettings.js]: Might make sense to pass in params into the constructor so that you can construct a `UserEmail` without having to know which fields are observables and set those observables manually (in `UserProfile#unserialize`).

Comment (sloria) [inline website/static/js/accountSettings.js]:
```js
constructor: function(params) {
    this.address = ko.observable(params.address || '');
    //...
}

File: website/static/js/accountSettings.js [hunk 31]
```diff
@@ -0,0 +1,187 @@
...
+        this.address = ko.observable();
+        this.isConfirmed = ko.observable();
+        this.isPrimary = ko.observable();
+    }
+});
+
+
+var UserProfile = oop.defclass({
+    constructor: function () {
```
Comment (sloria) [inline website/static/js/accountSettings.js]: See previous comment re: passing params.

File: website/templates/profile/account.mako [hunk 32]
```diff
@@ -5,39 +5,101 @@
...
+                                <tr>
+                                    <td style="width:100%">{{ $data.address }}</td>
+                                    <td><a data-bind="click: $parent.removeEmail"><i class="fa fa-times text-danger"></i></a></td>
+                                </tr>
+                                <!-- /ko -->
+                                <tr>
+                                    <td colspan="2">
+                                        <form data-bind="submit: addEmail">
+                                            <div class="form-group">
+                                                <input data-bind="value: emailInput" class="form-control" name="addEmail">
```
Comment (sloria) [inline website/templates/profile/account.mako]: The `name` ofr this input is camelCase while the others are snake_case.

File: website/templates/profile/account.mako [hunk 33]
```diff
@@ -5,39 +5,101 @@
...
+                        <table class="table">
+                            <thead>
+                                <tr>
+                                    <th colspan="3">Alternate Emails</th>
+                                </tr>
+                            </thead>
+                            <tbody data-bind="foreach: profile().alternateEmails()">
+                                <tr>
+                                    <td style="width:100%">{{ $data.address }}</td>
+                                    <td><a data-bind="click: $parent.makeEmailPrimary">make&nbsp;primary</a></td>
```
Comment (sloria) [inline website/templates/profile/account.mako]: Uppercase "Make"?

File: website/profile/views.py (hunk 16 above)
Comment (sloria) [inline website/profile/views.py]: Also, whitespace might need to be stripped.

**Comment by sloria** [comment]
2nd pass finished. :v:

File: tests/test_models_user.py (hunk 15 above)
Comment (lyndsysimon) [inline tests/test_models_user.py]: My intention was to move the User model tests there in time, yes. The `test_models.py` is huge and was slowing down my testing cycle, even just to parse it for the tests I wanted.

I like the idea of `test_models` (or, `/tests/models/user.py` if nose would detect the test), but I didn't want to make that big a change without consulting others.

I'll go ahead and move this when I get a chance.

File: website/profile/views.py (hunk 16 above)
Comment (lyndsysimon) [inline website/profile/views.py]: Good idea. Done.

File: website/profile/views.py (hunk 17 above)
Comment (lyndsysimon) [inline website/profile/views.py]: done

File: website/profile/views.py (hunk 18 above)
Comment (lyndsysimon) [inline website/profile/views.py]: I was looking for that!

I changed it; it was raising `ValueError` on validation failure. When I refactored to a separate validator, I changed it to raise `ValidationError` instead. I knew I was catching it somewhere, but couldn't find where. Thanks! :bow:

File: website/profile/views.py (hunk 19 above)
Comment (lyndsysimon) [inline website/profile/views.py]: Agreed. I didn't much like that structure the way it was before - this way lets me space it out a bit and make it more clear what's happening.

File: website/profile/views.py (hunk 20 above)
Comment (lyndsysimon) [inline website/profile/views.py]: Yes, but that's outside the scope of this PR and I don't want to add it right now.

I'll be working in this method extensively next week; I've added a TODO and will take care of it then.

File: website/static/js/accountSettings.js (hunk 21 above)
Comment (lyndsysimon) [inline website/static/js/accountSettings.js]: Are there guidelines yet on when to use a `pureComputed` vs. `computed`? I've read the Knockout docs and I think I grok when to use it, but I'd like to bounce it off you in person before I'm confident of my understanding.

File: website/static/js/accountSettings.js (hunk 22 above)
Comment (lyndsysimon) [inline website/static/js/accountSettings.js]: done

File: website/static/js/accountSettings.js (hunk 23 above)
Comment (lyndsysimon) [inline website/static/js/accountSettings.js]: done

File: website/static/js/accountSettings.js (hunk 24 above)
Comment (lyndsysimon) [inline website/static/js/accountSettings.js]: Talked it over, was a misunderstanding :)

File: website/static/js/accountSettings.js (hunk 25 above)
Comment (lyndsysimon) [inline website/static/js/accountSettings.js]: Leaving as-is for now; I need to do some thinking on style. It should be very easy to change in the future if need be.

File: framework/auth/core.py (hunk 27 above)
Comment (lyndsysimon) [inline framework/auth/core.py]: Yes, but we'll still need to call `update_search_nodes()` manually, since that's not checked upon save.

I've added a note to add that check :)

File: tests/test_models_user.py (hunk 15 above)
Comment (lyndsysimon) [inline tests/test_models_user.py]: I moved the file and have been moving tests that break there when I review them.

File: framework/auth/core.py (hunk 29 above)
Comment (lyndsysimon) [inline framework/auth/core.py]: Nothing should be in this field except lower-cased emails.

I've made a note about this; RFC 822 actually says that the `local-part` of an email - the part before the "@" - is case sensitive. Following that could lead to issues on our end, though. It bears further consideration.

Comment (GaryKriebel) [inline framework/auth/core.py]: I would think we would want to stay with insensitive to stay away from those issues.  Why would the standard go with case sensitivity?

File: website/templates/profile/account.mako (hunk 32 above)
Comment (lyndsysimon) [inline website/templates/profile/account.mako]: Removed the name entirely.

File: website/templates/profile/account.mako (hunk 33 above)
Comment (lyndsysimon) [inline website/templates/profile/account.mako]: I have no strong preference. It looked odd to me to capitalize "Make" since it isn't a sentence, and capitalizing "Primary" makes it look like a header and draw too much attention.

**Comment by lyndsysimon** [comment]
Bah. For some reason, Travis isn't building this every push.

I've got my repo to build as well, and this passes: https://travis-ci.org/lyndsysimon/osf.io/builds/56771321

**Comment by lyndsysimon** [comment]
Response to 2nd pass complete. Your move, @sloria...

:game_die:

File: framework/auth/core.py [hunk 34]
```diff
@@ -541,14 +596,47 @@ def _set_email_token_expiration(self, token, expiration=None):
...
+                del self.email_verifications[token]
+                return True
+
+        return False
+
+    def remove_email(self, email):
+        """Remove a confirmed email"""
+        if email in self.emails:
+            self.emails.remove(email)
+            mails.send_mail(self.username,
```
Comment (sloria) [inline framework/auth/core.py]: Might make sense to send these emails in a signal subscriber. You should be able to make a method a subscriber:

```python
    def remove_email(self):
        self.email.remove(email)
        user_email_removed.send(self, email)

    @user_email_removed.connect
    def send_email_removal_confirmations(self, email):
        if email in self.emails:
            #...
```

File: framework/auth/core.py (hunk 26 above)
Comment (sloria) [inline framework/auth/core.py]: I still think this implicit save can be removed.

File: framework/auth/core.py [hunk 35]
```diff
@@ -888,15 +1065,24 @@ def merge_user(self, user, save=False):
...
+        user.password = None
+        user.email_verifications = {}
+        user.verification_key = None
         user.merged_by = self
+
         user.save()
-        if save:
-            self.save()
-        return None
+        self.save()
```
Comment (sloria) [inline framework/auth/core.py]: Same here; I thought we decided to remove these?

File: website/profile/views.py [hunk 36]
```diff
@@ -85,20 +87,89 @@ def update_user(auth):
...
+        if primary_email:
+            username = primary_email['address'].strip().lower()
+
+        # make sure the new username has already been confirmed
+        if username and username in user.emails and username != user.username:
+            mails.send_mail(user.username,
+                            mails.PRIMARY_EMAIL_CHANGED,
+                            user=user,
+                            new_address=username)
+            user.username = username
```
Comment (sloria) [inline website/profile/views.py]: I wonder if it would make sense to have a `set_primary_email` method on User which

- Preprocesses email (lowercas and strip)
- Sets username field
- Emits a signal

Another method could subscribe to the signal and send the email.

File: website/templates/profile/account.mako (hunk 33 above)
Comment (sloria) [inline website/templates/profile/account.mako]: OK, it's fine as is.

**Comment by sloria** [comment]
Pass finished. :tanabata_tree:

File: framework/auth/core.py (hunk 34 above)
Comment (lyndsysimon) [inline framework/auth/core.py]: Done.

File: framework/auth/core.py (hunk 35 above)
Comment (lyndsysimon) [inline framework/auth/core.py]: Done.

I've also made a note to consider alternatives to this, where side effects - particularly time-intensive side effects - might be deferred until the request is complete, regardless of when or how often `StoredObject.save()` is called.

File: website/profile/views.py (hunk 36 above)
Comment (lyndsysimon) [inline website/profile/views.py]: Yes, I think that does make sense, but I'd like to defer it for the time being. With the upcoming changes to the API for `User`, this view will need to be refactored (or dynamically generated?) soon anyhow.

**Comment by lyndsysimon** [comment]
Third pass review complete.

:pager:ing @sloria.

File: website/static/js/accountSettings.js [hunk 37]
```diff
@@ -0,0 +1,190 @@
...
+        this.profile().emails.push(email);
+
+        this.client.update(this.profile()).done(function (profile) {
+            this.profile(profile);
+
+            var emails = profile.emails();
+            for (var i=0; i<emails.length; i++) {
+                if (emails[i].address() === email.address()) {
+                    this.emailInput('');
+                    $osf.growl('Email Added', '<em>' + email.address()  + '<em>', 'success');
```
Comment (sloria) [inline website/static/js/accountSettings.js]: Might want to let the user know that they should check their email.

**Comment by sloria** [comment]
Is there any way to resend a confirmation email?

**Comment by sloria** [comment]
Minor "bug": When adding an unconfirmed that is already in the list of unconfirmed emails, the growl message says "Email added" when it should probably say "Email already added" or something similar.

File: website/templates/profile/account.mako [hunk 38]
```diff
@@ -5,39 +5,101 @@
...
+                                <tr>
+                                    <td style="width:100%">{{ $data.address }}</td>
+                                    <td><a data-bind="click: $parent.removeEmail"><i class="fa fa-times text-danger"></i></a></td>
+                                </tr>
+                                <!-- /ko -->
+                                <tr>
+                                    <td colspan="2">
+                                        <form data-bind="submit: addEmail">
+                                            <div class="form-group">
+                                                <input data-bind="value: emailInput" class="form-control">
```
Comment (sloria) [inline website/templates/profile/account.mako]: You can take advantage of the HTML5's built-in validation by adding `type="email"`.

Also, might be helpful to add a placeholder, like `placeholder="New email"`.

**Comment by sloria** [comment]
Bug: Whenever I input an unconfirmed email with an uppercase character, I get an error message "Email validation failed", but the email still gets added.

File: website/templates/profile/account.mako [hunk 39]
```diff
@@ -5,39 +5,101 @@
...
+                                <input type="password" class="form-control" name="confirm_password">
+                            </div>
+                            <button type="submit" class="btn btn-default">Update password</button>
+                        </form>
+                    </div>
                 </div>
             </div>
         </div>
     </div>
+    <script src=${"/static/public/js/profile-account-settings-page.js" | webpack_asset}></script>
```
Comment (sloria) [inline website/templates/profile/account.mako]: This should be in the `def javascript_bottom` block.

File: website/static/js/pages/profile-account-settings-page.js (hunk 11 above)
Comment (sloria) [inline website/static/js/pages/profile-account-settings-page.js]: I still see the `.js` extensions?

File: framework/auth/core.py [hunk 40]
```diff
@@ -93,6 +94,15 @@ def validate_personal_site(value):
 def validate_social(value):
     validate_personal_site(value.get('personal'))
 
+
+def validate_email(item):
```
Comment (sloria) [inline framework/auth/core.py]: Possibly related to the bug I reported: this email validator is probably incomplete. Could use something like https://github.com/marshmallow-code/marshmallow/blob/c781939daaee64bf55ffc73d893bcb7c67aeba0d/marshmallow/validate.py#L69-122, which is a simplification of Django's email validator: https://github.com/django/django/blob/master/django/core/validators.py#L137-L205

File: website/templates/profile/account.mako [hunk 41]
```diff
@@ -5,39 +5,101 @@
...
-    <div class="row">
-        <div class="col-md-3">
-            <div class="panel panel-default">
-                <ul class="nav nav-stacked nav-pills">
-                    <li><a href="${ web_url_for('user_profile') }">Profile Information</a></li>
-                    <li><a href="#">Account Settings</a></li>
-                    <li><a href="${ web_url_for('user_addons') }">Configure Add-ons</a></li>
-                    <li><a href="${ web_url_for('user_notifications') }">Notifications</a></li>
-                </ul>
+    <div id="accountSettings">
```
Comment (sloria) [inline website/templates/profile/account.mako]: You can add the `scripted` CSS class to hide this div until the view model is bound; this prevents a Flash of Unstyled Content.

**Comment by sloria** [comment]
Doesn't have to be added before this is merged, but we might want to add a little helper text (e.g. with a question-mark button) to explain what it means to make an email "primary".

**Comment by sloria** [comment]
Pass finished. :deciduous_tree:

File: scripts/migration/migrate_unconfirmed_user_emails.py [hunk 42]
```diff
@@ -0,0 +1,45 @@
+"""Removes User.username from User.emails for unconfirmed users.
+
+"""
+
+import logging
```
Comment (sloria) [inline scripts/migration/migrate_unconfirmed_user_emails.py]: Missing import of `sys`.

**Comment by sloria** [comment]
Some input from Trello re: the "Email added" message: https://trello.com/c/X8aJ9lC7/73-more-information-when-email-is-added

**Comment by sloria** [comment]
Also: https://trello.com/c/Udvre4jB/75-adding-an-email-for-a-confirmed-account-with-no-email-set

File: framework/auth/core.py (hunk 40 above)
Comment (lyndsysimon) [inline framework/auth/core.py]: I'm OK with it as-is if you are. It's as minimal as possible, really only checking to see if it "looks like" an email. I figure the only way to verify that it *is* an email is to confirm it, which we're doing :)

File: scripts/migration/migrate_unconfirmed_user_emails.py (hunk 42 above)
Comment (lyndsysimon) [inline scripts/migration/migrate_unconfirmed_user_emails.py]: Fixed

**Comment by chennan47** [comment]
ed20e52 address https://trello.com/c/X8aJ9lC7/73-more-information-when-email-is-added

**Comment by chennan47** [comment]
@sloria rdy for review

**Comment by lbanner** [comment]
:+1:
